Partial Sync UA stylesheet rules as per HTML Spec for 'lists'

Ahmad-S792 opened the pull request:
#### 5febb62e57ba1916cc54cc3cfafc7ac5aef9f003
<pre>
Partial Sync UA stylesheet rules as per HTML Spec for &apos;lists&apos;

<a href="https://bugs.webkit.org/show_bug.cgi?id=253177">https://bugs.webkit.org/show_bug.cgi?id=253177</a>

Reviewed by Tim Nguyen.

This patch aligns WebKit with web-spec [1] and remove quirks or additions:

[1] <a href="https://html.spec.whatwg.org/#lists">https://html.spec.whatwg.org/#lists</a>

NOTE: We are still using &apos;1__qem&apos; for `margin-block-start`
while we are not adding &apos;counter-reset: list-item&apos; for `ol`.

* Source/WebCore/css/html.css: Update &apos;lists&apos; section
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/lists/lists-styles-expected.txt: Rebaselined
* LayoutTests/platform/glib/fast/lists/009-expected.txt: Rebaselined
* LayoutTests/platform/glib/fast/lists/009-vertical-expected.txt: Rebaselined
* LayoutTests/platform/glib/fast/parser/001-expected.txt: Rebaselined
* LayoutTests/platform/gtk/fast/lists/009-expected.png: Rebaselined
* LayoutTests/platform/gtk/fast/lists/009-vertical-expected.png: Rebaselined
* LayoutTests/platform/gtk/fast/parser/001-expected.png: Rebaselined
* LayoutTests/platform/ios/fast/lists/009-expected.txt: Rebaselined
* LayoutTests/platform/ios/fast/lists/009-vertical-expected.txt: Rebaselined
* LayoutTests/platform/ios/fast/parser/001-expected.txt: Rebaselined
* LayoutTests/platform/mac/fast/parser/001-expected.txt: Rebaselined
* LayoutTests/platform/mac/fast/parser/001-expected.png: Rebaselined
* LayoutTests/platform/mac/fast/lists/009-vertical-expected.txt: Rebaselined
* LayoutTests/platform/mac/fast/lists/009-vertical-expected.png: Rebaselined
* LayoutTests/platform/mac/fast/lists/009-expected.txt: Rebaselined
* LayoutTests/platform/mac/fast/lists/009-expected.png: Rebaselined

Canonical link: <a href="https://commits.webkit.org/264052@main">https://commits.webkit.org/264052@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/af918962c00828e84742a2447329da7b797aa626

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6556 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6768 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6949 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8138 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6841 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6553 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/7294 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6718 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9736 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6671 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/6853 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5961 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8226 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/4310 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5925 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13759 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/6283 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6005 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/8435 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6489 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5311 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5890 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1542 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10055 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6262 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/html.css:
```diff
@@ -304,43 +304,45 @@ caption {
     display: block;
     list-style-type: decimal;
-    margin-block-start: 1__qem;
+    counter-reset: list-item;
```
This and the other `counter-reset: list-item` rule also need the following part of the spec implemented:

<img width="1436" alt="image" src="https://github.com/WebKit/WebKit/assets/1795135/42f3bc8a-a309-4789-bb72-19f432c3c466">

otherwise you'll regress `imported/w3c/web-platform-tests/css/css-lists/counter-list-item.html`

nt1m replied:
It's not super hard to implement if you want to try, the relevant code is here:

https://searchfox.org/wubkat/rev/e4cb3bdcaddcf79ba58dab47b0245b781c7db975/Source/WebCore/html/HTMLLIElement.cpp#66

https://searchfox.org/wubkat/rev/e4cb3bdcaddcf79ba58dab47b0245b781c7db975/Source/WebCore/html/HTMLOListElement.cpp#70

Though there's a bit of cleanup to do (happy to show you what cleanup), since there's also some pre-existing code that does similar things.

nt1m replied:
I take that back, it's not trivial to implement at all because WebKit doesn't support `reversed()` counters.

Ahmad-S792 replied:
Removed this for time being and updated commit message as well.

nt1m commented on Source/WebCore/css/html.css:
```diff
@@ -304,43 +304,45 @@ caption {
 ul, menu, dir {
     display: block;
     list-style-type: disc;
-    margin-block-start: 1__qem;
```
I'd rather do this change in a separate bug, so it can have enough discussion. This quirk has been around for a long time (since 2002? or even earlier): https://github.com/WebKit/WebKit/commit/ad72cdff4371a3c296ab7a7e162620ebdc940325

Ahmad-S792 replied:
Reversed for time being and raised new [bug](https://bugs.webkit.org/show_bug.cgi?id=256754).

nt1m commented on Source/WebCore/css/html.css:
```diff
@@ -316,31 +314,34 @@ ol {
+    margin-block-start: 1em;
+    margin-block-end: 1em;
+}
+
```
Can we remove this block? It's covered by the previous rule, with the difference that the previous rule is using __qem

Ahmad-S792 replied:
Good spot!

nt1m commented on Source/WebCore/css/html.css:
```diff
@@ -316,31 +314,34 @@ ol {
-    text-transform: none;
+ul, menu {
+    counter-reset: list-item;
 }
```
This will break the list-item counter without the presentational hints implemented. Can we undo this too?

Ahmad-S792 replied:
Done.

nt1m reviewed (commented):
Looks mostly good, thanks!

Ahmad-S792 committed:
Partial Sync UA stylesheet rules as per HTML Spec for 'lists'

https://bugs.webkit.org/show_bug.cgi?id=253177

Reviewed by Tim Nguyen.

This patch aligns WebKit with web-spec [1] and remove quirks or additions:

[1] https://html.spec.whatwg.org/#lists

NOTE: We are still using '1__qem' for `margin-block-start`
while we are not adding 'counter-reset: list-item' for `ol`.

* Source/WebCore/css/html.css: Update 'lists' section
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/lists/lists-styles-expected.txt: Rebaselined
* LayoutTests/platform/glib/fast/lists/009-expected.txt: Rebaselined
* LayoutTests/platform/glib/fast/lists/009-vertical-expected.txt: Rebaselined
* LayoutTests/platform/glib/fast/parser/001-expected.txt: Rebaselined
* LayoutTests/platform/gtk/fast/lists/009-expected.png: Rebaselined
* LayoutTests/platform/gtk/fast/lists/009-vertical-expected.png: Rebaselined
* LayoutTests/platform/gtk/fast/parser/001-expected.png: Rebaselined
* LayoutTests/platform/ios/fast/lists/009-expected.txt: Rebaselined
* LayoutTests/platform/ios/fast/lists/009-vertical-expected.txt: Rebaselined
* LayoutTests/platform/ios/fast/parser/001-expected.txt: Rebaselined
* LayoutTests/platform/mac/fast/parser/001-expected.txt: Rebaselined
* LayoutTests/platform/mac/fast/parser/001-expected.png: Rebaselined
* LayoutTests/platform/mac/fast/lists/009-vertical-expected.txt: Rebaselined
* LayoutTests/platform/mac/fast/lists/009-vertical-expected.png: Rebaselined
* LayoutTests/platform/mac/fast/lists/009-expected.txt: Rebaselined
* LayoutTests/platform/mac/fast/lists/009-expected.png: Rebaselined

Canonical link: https://commits.webkit.org/264052@main
